### PR TITLE
implement Autofill Transaction

### PR DESCRIPTION
**Pull request by khancode**
## High Level Overview of Change

Implements `autofill()` and `setTransactionFlagsToNumber()` to allow a `Client` to autofill fields in a `Transaction`.

### Context of Change
`autofill()` does the following operations:
- Converts X-address to classic address for each account field.
- Calls `setTransactionFlagsToNumber()` that sets a transaction's flags (object of boolean flags) to its numeric representation.
- Autofills missing `Transaction` fields: `Fee`, `Sequence`, `LastLedgerSequence`.

### Type of Change

- [x] New feature (non-breaking change which adds functionality)
- [x] Tests (You added tests for code that already exists, or your new feature included in this PR)

## Test Plan

Adds tests for `autofill()` and `setTransactionFlagsToNumber()`

### CODE REVIEW

File: src/client/index.ts [hunk 1]
```diff
@@ -338,6 +339,16 @@ class Client extends EventEmitter {
     return transactionUtils.convertStringToHex(string);
   }
 
+  /**
+   * Autofills missing fields in a transaction
+   *
+   * @param {Transaction} tx A transaction to autofill missing fields
+   * @returns {Transaction} An autofilled transaction
```
Comment (mvadari) [inline src/client/index.ts:346]: nit: we're using a different doc format that the linter is enforcing

File: src/models/transactions/common.ts [hunk 2]
```diff
@@ -66,6 +66,10 @@ export function isAmount(amount: Amount): boolean {
   return typeof amount === "string" || isIssuedCurrency(amount);
 }
 
+export enum GlobalFlagsEnum {
+  tfFullyCanonicalSig = 0x80000000,
+}
+
```
Comment (mvadari) [inline src/models/transactions/common.ts:72]: This flag isn't actually used anymore: https://xrpl.org/transaction-common-fields.html#global-flags

File: src/models/utils/index.ts [hunk 3]
```diff
@@ -19,3 +22,129 @@ export function onlyHasFields(obj: object, fields: string[]): boolean {
...
+      tx.Flags = convertPaymentChannelClaimFlagsToNumber(tx.Flags as PaymentChannelClaimFlags);
+      break;
+    case 'Payment':
+      tx.Flags = convertPaymentTransactionFlagsToNumber(tx.Flags as PaymentTransactionFlags);
+      break;
+    case 'TrustSet':
+      tx.Flags = convertTrustSetFlagsToNumber(tx.Flags as TrustSetFlags);
+      break;
+    default:
+      tx.Flags = convertGlobalFlagsToNumber(tx.Flags);
```
Comment (mvadari) [inline src/models/utils/index.ts:47]: `convertGlobalFlagsToNumber` should be applied on all transactions, regardless of their type

File: src/models/utils/index.ts [hunk 4]
```diff
@@ -1,3 +1,6 @@
+import { OfferCreateFlags, OfferCreateFlagsEnum, PaymentChannelClaimFlags, PaymentChannelClaimFlagsEnum, PaymentTransactionFlags, PaymentTransactionFlagsEnum, Transaction, TrustSetFlags, TrustSetFlagsEnum } from "../transactions";
```
Comment (mvadari) [inline src/models/utils/index.ts:1]: Run `eslint --fix` on the new code you're adding

File: src/transaction/utils.ts [hunk 5]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
+ * @param {Transaction} tx A transaction to autofill fields.
+ * @param {Client} client A client.
+ * @returns {Transaction} An autofilled transaction.
+ */
+async function autofillTransaction(
+  tx: Transaction,
+  client: Client
+): Promise<Transaction> {
+  validateAccountAddress(tx, 'Account', 'SourceTag');
+  if (tx['Destination']) {
```
Comment (mvadari) [inline src/transaction/utils.ts:432]:
```suggestion
  if (tx['Destination'] == null) {
```

File: src/transaction/utils.ts [hunk 6]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
+  convertToClassicAddress(tx, 'Authorize');
+  convertToClassicAddress(tx, 'Unauthorize');
+  // EscrowCancel, EscrowFinish:
+  convertToClassicAddress(tx, 'Owner');
+  // SetRegularKey:
+  convertToClassicAddress(tx, 'RegularKey');
+
+  setTransactionFlagsToNumber(tx);
+
+  if (tx.Sequence === undefined) {
```
Comment (mvadari) [inline src/transaction/utils.ts:446]:
```suggestion
  if (tx.Sequence == null) {
```
We want to avoid using `undefined` to be slightly more general

File: src/transaction/utils.ts [hunk 7]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
+    const sequence = await getNextValidSequenceNumber(tx.Account, client);
+    tx.Sequence = sequence;
+  }
+
+  if (tx.Fee === undefined) {
```
Comment (mvadari) [inline src/transaction/utils.ts:451]: ditto

File: src/transaction/utils.ts [hunk 8]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
+    tx.Fee = await calculateFeePerTransactionType(tx, client);
+  }
+
+  if (tx.LastLedgerSequence === undefined) {
```
Comment (mvadari) [inline src/transaction/utils.ts:455]: ditto

File: src/transaction/utils.ts [hunk 9]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
+    account,
+  };
+  const data = await client.request(request);
+  return data.result.account_data.Sequence;
+}
+
+async function calculateFeePerTransactionType(
+  transaction: Transaction,
+  client?: Client
+): Promise<string> {
```
Comment (mvadari) [inline src/transaction/utils.ts:501]: Something that's different about xrpl-py vs xrpl.js (which needs to be changed in Python) is that you'll definitely have a client for this. There should not be a default fee.

File: src/transaction/utils.ts [hunk 10]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
+  let netFee;
+  if (client === undefined) {
+    netFee = 10; // 10 drops
+  } else {
+    netFee = await client.getFee(); // Usually 0.00001 XRP (10 drops)
+  }
+
+  let baseFee = netFee;
```
Comment (mvadari) [inline src/transaction/utils.ts:509]: You might need to do some more processing on this - https://github.com/ripple/ripple-lib/blob/develop/src/transaction/utils.ts#L278

**Review by mvadari — CHANGES_REQUESTED** [review]

File: src/transaction/utils.ts (hunk 5 above)
Comment (khancode) [reply src/transaction/utils.ts:432]: This should be `!= null` instead since it's checking if Destination is defined.

File: src/models/transactions/common.ts (hunk 2 above)
Comment (khancode) [reply src/models/transactions/common.ts:72]: Since this is deprecated and the only flag in `GlobalFlags`, we should remove `GlobalFlags` since it'll become an empty interface. @natenichols defined the `GlobalFlags` interface in another [commit](https://github.com/ripple/ripple-lib/commit/3231c9e08c6ff469876d5ff99093e32a72f12852). How do you feel about my suggestion?

Comment (khancode) [reply src/models/transactions/common.ts:72]: Since this is deprecated and the only flag in GlobalFlags, we should remove GlobalFlags since it'll become an empty interface. @natenichols defined the GlobalFlags interface in another commit. What do you think about my suggestion?

File: src/models/utils/index.ts (hunk 3 above)
Comment (khancode) [reply src/models/utils/index.ts:47]: It technically is in my PR but looks like it's not apparent reading it. I call `convertGlobalFlagsToNumber` in each `convert{TransactionType}FlagsToNumber` so that these functions return the full numeric value including GlobalFlags.

If we still want to keep `GlobalFlags` interface, I can move `convertGlobalFlagsToNumber` to the top of `setTransactionFlagsToNumber` and specify `convert{TransactionType}FlagsToNumber` exclude `GlobalFlags`.

Comment (khancode) [reply src/models/utils/index.ts:47]: It technically is in my PR but looks like it's not apparent reading it. I call `convertGlobalFlagsToNumber` in each `convert{TransactionType}FlagsToNumber` so that these functions return the full numeric value including `GlobalFlags`.

If we still want to keep `GlobalFlags` interface, I can move `convertGlobalFlagsToNumber` to the top of `setTransactionFlagsToNumber` and specify `convert{TransactionType}FlagsToNumber` exclude `GlobalFlags`.

Comment (khancode) [reply src/models/utils/index.ts:47]: If we still want to keep `GlobalFlags` interface, I can move `convertGlobalFlagsToNumber` to the top of `setTransactionFlagsToNumber` and specify `convert{TransactionType}FlagsToNumber` exclude `GlobalFlags`.

File: src/client/index.ts (hunk 1 above)
Comment (natenichols) [reply src/client/index.ts:346]: This should be resolvable with a `yarn lint --fix`

File: src/models/transactions/common.ts (hunk 2 above)
Comment (natenichols) [reply src/models/transactions/common.ts:72]: We need to have a `GlobalFlags`, because otherwise we'd have to specify flags in the baseTransaction as `Flags: number | Record<string,unkown>` We can do that, but i think its a lot cleaner to have `GlobalFlags` serve as a base class.

File: src/transaction/utils.ts [hunk 11]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
 
+/**
+ * Autofills fields in a transaction.
+ *
+ * @param {Transaction} tx A transaction to autofill fields.
+ * @param {Client} client A client.
+ * @returns {Transaction} An autofilled transaction.
+ */
+async function autofillTransaction(
+  tx: Transaction,
```
Comment (natenichols) [inline src/transaction/utils.ts:428]: We'll want this on the client, so I think we'll want to define this as 

```ts
async function autofillTransaction(
    this: Client,
    tx: Transaction
```
and then 
```ts
class Client {
    autofillTransaction = autofillTransaction
}

**Review by natenichols — APPROVED** [review]
LGTM 👍

File: src/client/index.ts [hunk 12]
```diff
@@ -338,6 +339,16 @@ class Client extends EventEmitter {
...
+   */
+  async autofillTransaction(tx: Transaction): Promise<Transaction> {
+    return transactionUtils.autofillTransaction(tx, this)
+  }
```
Comment (ledhed2222) [inline src/client/index.ts:350]: why not call this `autofill`? the signature will make it clear that we expect a transaction, and there aren't other things to autofill. Also, shouldn't you `await` the function from `transactionUtils`? Otherwise you could instead write this as:

```typescript
import { partial } from 'lodash'

autofillTransaction = partial(transactionUtils.autofillTransaction, _, this)
```

File: src/models/utils/index.ts (hunk 4 above)
Comment (ledhed2222) [reply src/models/utils/index.ts:1]: we'll get to this later - since the linter rules aren't there if i were you omar I would wait actually

File: src/models/utils/index.ts [hunk 13]
```diff
@@ -19,3 +22,129 @@ export function onlyHasFields(obj: object, fields: string[]): boolean {
...
+ * Sets a transaction's flags to its numeric representation
+ *
+ * @param {Transaction} tx A transaction to set its flags to its numeric representation
+ * @returns {void}
+ */
+export function setTransactionFlagsToNumber(tx: Transaction): void {
+  switch (tx.TransactionType) {
+    case 'OfferCreate':
+      tx.Flags = convertOfferCreateFlagsToNumber(tx.Flags as OfferCreateFlags);
+      break;
```
Comment (ledhed2222) [inline src/models/utils/index.ts:36]: maybe `return` instead of `break`. no real reason other than that the matched expression should be the last expression for the function anyway, so i think that that makes it more clear to the reader, who knows there is nothing below the `switch` statement

File: src/models/utils/index.ts [hunk 14]
```diff
@@ -19,3 +22,129 @@ export function onlyHasFields(obj: object, fields: string[]): boolean {
...
+  if (typeof flags === 'number') {
+    return flags;
+  }
+  return flags?.tfFullyCanonicalSig ? GlobalFlagsEnum.tfFullyCanonicalSig : 0;
+}
+
+function convertOfferCreateFlagsToNumber(flags: OfferCreateFlags): number {
+  if (flags == null) {
+    return 0;
+  } else if (typeof flags === 'number') {
```
Comment (ledhed2222) [inline src/models/utils/index.ts:61]: this is a personal pet peeve of mine for some reason, but the `else` is unnecessary here since you returned in the first conditional

```typescript
if (flags == null) {
  return 0
}
if (typeof flags === 'number') {
  return flags
}
....
```

File: src/models/utils/index.ts [hunk 15]
```diff
@@ -19,3 +22,129 @@ export function onlyHasFields(obj: object, fields: string[]): boolean {
...
+  }
+  if (flags.tfImmediateOrCancel) {
+    resultFlags |= OfferCreateFlagsEnum.tfImmediateOrCancel;
+  }
+  if (flags.tfFillOrKill) {
+    resultFlags |= OfferCreateFlagsEnum.tfFillOrKill;
+  }
+  if (flags.tfSell) {
+    resultFlags |= OfferCreateFlagsEnum.tfSell;
+  }
```
Comment (ledhed2222) [inline src/models/utils/index.ts:78]: seems like there is a way to express this as a `reduce` instead of a big series of conditions. not a blocker but i think that would be more elegant

File: src/models/utils/index.ts [hunk 16]
```diff
@@ -19,3 +22,129 @@ export function onlyHasFields(obj: object, fields: string[]): boolean {
...
+
+  return resultFlags;
+}
+
+function convertPaymentChannelClaimFlagsToNumber(flags: PaymentChannelClaimFlags): number {
+  if (flags == null) {
+    return 0;
+  } else if (typeof flags === 'number') {
```
Comment (ledhed2222) [inline src/models/utils/index.ts:86]: as above

File: src/models/utils/index.ts [hunk 17]
```diff
@@ -19,3 +22,129 @@ export function onlyHasFields(obj: object, fields: string[]): boolean {
...
+
+  return resultFlags;
+}
+
+function convertPaymentTransactionFlagsToNumber(flags: PaymentTransactionFlags): number {
+  if (flags == null) {
+    return 0;
+  } else if (typeof flags === 'number') {
+    return flags;
+  }
```
Comment (ledhed2222) [inline src/models/utils/index.ts:107]: this is now repeated a few times. i think you should factor this out into its own function

File: src/transaction/utils.ts (hunk 11 above)
Comment (ledhed2222) [reply src/transaction/utils.ts:428]: also agree that the `client` should be the first argument, as that makes usage with partial application (see above) or currying easier

File: src/transaction/utils.ts (hunk 9 above)
Comment (ledhed2222) [reply src/transaction/utils.ts:501]: i would also again make the client the first arg

File: src/transaction/utils.ts [hunk 18]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
+  const account = tx[fieldName];
+  if (typeof account === "string") {
+    const { classicAccount } = getClassicAccountAndTag(account);
+    tx[fieldName] = classicAccount;
+  }
+}
+
+async function getNextValidSequenceNumber(
+  account: string,
+  client: Client
```
Comment (ledhed2222) [inline src/transaction/utils.ts:488]: client shoudl be the first arg

File: src/transaction/utils.ts [hunk 19]
```diff
@@ -412,7 +417,136 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
+
+  // EscrowFinish Transaction with Fulfillment
+  if (transaction.TransactionType === 'EscrowFinish') {
+    if (transaction.Fulfillment != null) {
+      const fulfillmentBytesSize = Math.ceil(
+        transaction.Fulfillment.length / 2
+      );
+      // 10 drops × (33 + (Fulfillment size in bytes / 16))
+      baseFee = Math.ceil(netFee * (33 + fulfillmentBytesSize / 16));
+    }
```
Comment (ledhed2222) [inline src/transaction/utils.ts:519]: this conditional nesting is unnecessary since there is nothing that happens if the outer condition is true but the inner condition is false. as a result you should unnest this. here's one suggestion

```typescript
const isFulfillment = transaction.TransactionType === 'EscrowFinish' && transaction.Fulfillment != null
if (isFullfillment) {
  const fulfillmentBytesSize = Math.ceil(.....)
  ....
}
```

File: src/transaction/utils.ts (hunk 11 above)
Comment (ledhed2222) [reply src/transaction/utils.ts:428]: i also think this should just be called `autofill` since it's already in the `transaction/utils.ts` file - that makes it clear what is being autofilled

**Review by ledhed2222 — CHANGES_REQUESTED** [review]
nice work - see comments

File: src/client/index.ts (hunk 12 above)
Comment (khancode) [reply src/client/index.ts:350]: I agree with your suggestions. However, I'm not a fan of importing a module when this can be done natively with JavaScript's `bind()`:
```typescript
autofill = transactionUtils.autofill.bind(null, this);
```
What do you think?

File: src/models/utils/index.ts (hunk 13 above)
Comment (khancode) [reply src/models/utils/index.ts:40]: Sure, makes sense.

File: src/models/utils/index.ts (hunk 14 above)
Comment (khancode) [reply src/models/utils/index.ts:65]: Haha, I personally like using `else if` so it's easy to tell that only one of the possible conditions will execute.

File: src/client/index.ts (hunk 12 above)
Comment (ledhed2222) [reply src/client/index.ts:350]: i'm no javascript expert, but in my opinion the weirdness of `this` in javascript is such that it should be avoided. but if you'd like to chat about this more i'm down! maybe this is the right move.

File: src/transaction/utils.ts (hunk 19 above)
Comment (khancode) [reply src/transaction/utils.ts:514]: Yup you're right, this totally went over my head copying our Python implementation. 😄

File: src/models/transactions/common.ts (hunk 2 above)
Comment (mvadari) [reply src/models/transactions/common.ts:101]: I'm fine with leaving it as an empty interface - it makes it easier to add other global flags if more get added in the future.

File: src/models/utils/index.ts (hunk 4 above)
Comment (mvadari) [reply src/models/utils/index.ts:1]: The linter rules are already in `2.0` courtesy of #1560

File: src/models/utils/index.ts [hunk 20]
```diff
@@ -23,3 +26,129 @@ export function isFlagEnabled(Flags: number, checkFlag: number): boolean {
   // eslint-disable-next-line no-bitwise -- Flags require bitwise operations
   return (checkFlag & Flags) === checkFlag;
 }
+
+/**
+ * Sets a transaction's flags to its numeric representation
+ *
+ * @param {Transaction} tx A transaction to set its flags to its numeric representation
+ * @returns {void}
+ */
```
Comment (mvadari) [inline src/models/utils/index.ts:35]: This isn't the right TSdocs format

File: src/models/utils/index.ts (hunk 3 above)
Comment (mvadari) [reply src/models/utils/index.ts:51]: I don't quite understand what you mean

Comment (khancode) [reply src/models/utils/index.ts]: I just refactored it, it should be clear now. 🙂

File: src/models/utils/index.ts (hunk 20 above)
Comment (khancode) [reply src/models/utils/index.ts]: I'll be fixing the eslint errors in another commit.

File: src/transaction/utils.ts [hunk 21]
```diff
@@ -412,7 +417,130 @@ function convertMemo(memo: Memo): { Memo: ApiMemo } {
...
 }
 
+/**
+ * Autofills fields in a transaction.
+ *
+ * @param client - A client.
+ * @param tx - A transaction to autofill fields.
+ * @returns An autofilled transaction.
+ */
+async function autofill(client: Client, tx: Transaction): Promise<Transaction> {
```
Comment (mvadari) [inline src/transaction/utils.ts:427]: This file should go in the `ledger` folder - something like `src/ledger/autofill.ts`
(yes, this doesn't really have to do anything with the ledger, but that's the folder all the sugar is going in, and it'll be renamed soon)

File: src/client/index.ts (hunk 12 above)
Comment (khancode) [reply src/client/index.ts]: @ledhed2222 and I further discussed this in a meeting and we agreed to create a sugar function for JavaScript `bind()`. These changes have been included.

File: src/ledger/autofill.ts [hunk 22]
```diff
@@ -0,0 +1,209 @@
...
+    tag: expectedTag,
+  };
+}
+
+/**
+ * Autofills fields in a transaction.
+ *
+ * @param client - A client.
+ * @param tx - A transaction to autofill fields.
+ * @param signersCount - A signers count used for multisign.
```
Comment (mvadari) [inline src/ledger/autofill.ts:68]:
```suggestion
 * @param signersCount - The expected number of signers for this transaction. Used for multisign.
```

File: src/ledger/autofill.ts [hunk 23]
```diff
@@ -0,0 +1,209 @@
...
+ * @returns An autofilled transaction.
+ */
+async function autofill(
```
Comment (mvadari) [inline src/ledger/autofill.ts:71]: The helper functions should all be above or below `autofill` - it's confusing when some are above and some are below

File: src/ledger/autofill.ts [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+    promises.push(setNextValidSequenceNumber(client, tx));
+  }
+  if (tx.Fee == null) {
+    promises.push(calculateFeePerTransactionType(client, tx, signersCount));
+  }
+  if (tx.LastLedgerSequence == null) {
+    promises.push(setLatestValidatedLedgerSequence(client, tx));
+  }
+
+  return Promise.all(promises).then(() => tx);
```
Comment (mvadari) [inline src/ledger/autofill.ts:91]: This is really nice

File: src/ledger/autofill.ts [hunk 25]
```diff
@@ -0,0 +1,209 @@
...
+ *
+ * @param client - A client.
+ * @param tx - A transaction to autofill fields.
+ * @param signersCount - A signers count used for multisign.
+ * @returns An autofilled transaction.
+ */
+async function autofill(
+  client: Client,
+  tx: Transaction,
+  signersCount?: number
```
Comment (mvadari) [inline src/ledger/autofill.ts:74]: @natenichols thoughts on an optional param for a `LastLedgerSequence` offset?

Comment (natenichols) [reply src/ledger/autofill.ts:74]: I would say leave it out IMO. You can accomplish the same thing w/ `LastLedgerSequence`. 

The big advantage to adding be that we could have a default`LastLedgerSequenceOffset`, such that clients would ensure that their transactions would succeed or fail within 10 ( or so ) ledgers.

Comment (mvadari) [reply src/ledger/autofill.ts:74]: There's already a default here, so I think we're good on that front. 
I guess we can always add that feature later.

File: src/client/index.ts [hunk 26]
```diff
@@ -20,9 +20,11 @@ import {
   decodeXAddress,
 } from "ripple-address-codec";
 
+import { bind } from "..";
```
Comment (ledhed2222) [inline src/client/index.ts:23]: why call this bind? the whole point i thought was to call it `partial`

Comment (khancode) [reply src/client/index.ts:23]: Ah you're right, slipped my mind!

File: src/ledger/autofill.ts (hunk 24 above)
Comment (khancode) [reply src/ledger/autofill.ts:51]: Thanks! 😊

File: src/client/index.ts [hunk 27]
```diff
@@ -1,7 +1,7 @@
 /* eslint-disable import/max-dependencies -- Client needs a lot of dependencies by definition */
 /* eslint-disable @typescript-eslint/member-ordering -- TODO: remove when instance methods aren't members */
 /* eslint-disable max-lines -- This might not be necessary later, but this file needs to be big right now */
-import { EventEmitter } from 'events'
```
Comment (mvadari) [inline src/client/index.ts:4]: Please incorporate prettier changes

File: src/client/index.ts [hunk 28]
```diff
@@ -198,6 +200,9 @@ class Client extends EventEmitter {
   // number. Defaults to '2'.
   public readonly maxFeeXRP: string
 
+  // eslint-disable-next-line @typescript-eslint/no-invalid-this -- this is expected
+  public autofill = partial(autofill, this)
```
Comment (mvadari) [inline src/client/index.ts:204]: can you add a TODO to use `partial` for the other methods as well? (you don't need to do it now, especially because we're going to delete a bunch of these methods anyways)

Comment (khancode) [reply src/client/index.ts:205]: Done.

File: src/utils/index.ts [hunk 29]
```diff
@@ -214,7 +214,21 @@ function ISOTimeToRippleTime(iso8601: string): number {
...
 
+/**
+ * It returns a function that prepends params to the given func.
+ * A sugar function for JavaScript .bind().
+ *
+ * @param func - A function to prepend params.
+ * @param params - Parameters to prepend to a function.
+ * @returns A function bound with params.
+ */
+function partial(func: Function, ...params: any[]): Function {
```
Comment (mvadari) [inline src/utils/index.ts:225]: Given that this is a helper method for the client, I'd recomment putting this in `src/client` so it doesn't need to be exported.

File: test/client/autofill.ts [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+    this.mockRippled.addResponse('ledger', {
+      status: 'success',
+      type: 'response',
+      result: {
+        ledger_index: 9038214,
+      },
+    })
+    const txResult = await this.client.autofill(tx)
+    assert.strictEqual(txResult.LastLedgerSequence, 9038234)
+  })
```
Comment (mvadari) [inline test/client/autofill.ts:201]: I would add one more test that checks if everything works all together (if all of the autofillable fields are empty, will it autofill all of them correctly), given that you're using a lot of promises.

File: test/models/utils.ts [hunk 31]
```diff
@@ -18,13 +32,124 @@ describe('Models Utils', function () {
...
+        Flags: {
+          tfPassive: true,
+          tfImmediateOrCancel: false,
+          tfFillOrKill: true,
+          tfSell: false,
+        },
+      }
+
+      const { tfPassive, tfFillOrKill } = OfferCreateFlagsEnum
+      // eslint-disable-next-line no-bitwise -- expected use of bitwise
```
Comment (mvadari) [inline test/models/utils.ts:69]: Just disable this for the whole file - it'll be easier/cleaner

**Review by mvadari — APPROVED** [review]
A few notes, nothing to hold up approval. Please address them before merging.

File: src/utils/index.ts (hunk 29 above)
Comment (khancode) [reply src/utils/index.ts:225]: I put it here because I see we use Lodash's version of `partial` in other files too. I'm planning on creating another PR to replace all Lodash partial with this one.

Comment (mvadari) [reply src/utils/index.ts:225]: Everywhere we're using it will be scrapped, so don't worry about that. You can even make the method tailored to the client.

File: src/client/index.ts [hunk 32]
```diff
@@ -169,6 +170,20 @@ function getCollectKeyFromCommand(command: string): string | null {
...
+/**
+ * It returns a function that prepends params to the given func.
+ * A sugar function for JavaScript .bind() without the "this" (keyword) binding.
+ *
+ * @param func - A function to prepend params.
+ * @param params - Parameters to prepend to a function.
+ * @returns A function bound with params.
+ */
+// eslint-disable-next-line @typescript-eslint/ban-types -- expected param types
+function prepend(func: Function, ...params: unknown[]): Function {
```
Comment (khancode) [inline src/client/index.ts:182]: @ledhed2222 Since `partial` will only be used in `Client` (@mvadari), I moved it here and renamed it to `prepend` to tailor it more to `Client` class.

**Review by ledhed2222 — APPROVED** [review]
